### ISSUE TITLE
Pyright failing to match recursive (nested?) generic protocol

### ISSUE THREAD

**Issue by timabroad** [microsoft/pyright#6183]
**Describe the bug**
I've got some code that defines a bunch of classes that are effectively representing custom types to schema a custom data format. So for example there an `Int` class, and `String`, and `Boolean`, and `List[ElemT]` etc. Each of these classes define a bunch of classmethods for doing stuff like parsing a value or formatting a value for output. For example, the `Int.convert()` function might be defined as `def convert(cls, value: str) -> int: ...`. I'm then using a generic Protocol to match these classes and in particular match against an "input" type and an "output" type. For example `Int` might match `TypeProto[str, int]` (by matching the `convert` function from above).

So far, so good, but now I want to be able to make `List[ElemT]` match `TypeProto[Iterable[InputElemT], list[OutputElemT]]` - that is, a `List[ElemT]` should match the protocol with input type `Iterable` and output type `list`, but correctly generically typed based on the actual type of `ElemT`, which itself matches the protocol. I've got this working with mypy, and almost working with pyright, but there seems to be something going wrong in pyright that's preventing it from quite working.

**Code or Screenshots**
```python
from typing import Any, Never, Generic, Self, TypeVar, Protocol, Iterable, reveal_type

# I've spelled these all out differently just to make it easy to see which one is which when trying to parse the pyright output.
In1T_contra = TypeVar('In1T_contra', contravariant = True)
In2T_contra = TypeVar('In2T_contra', contravariant = True)
In3T = TypeVar('In3T')
In4T = TypeVar('In4T')
Out1T = TypeVar('Out1T')
Out2T = TypeVar('Out2T')
Out3T = TypeVar('Out3T')
Out4T = TypeVar('Out4T') 
ElemT = TypeVar('ElemT')

# Cut down version of then protocol for matching my custom type classes.
class TypeProto(Protocol[In1T_contra, Out1T]):
    @classmethod
    def convert(cls, value: In1T_contra) -> Out1T:
        ...
    def resolve_item(self) -> 'TypeProto[In1T_contra, Out1T]':
        ...
    @classmethod
    def resolve_type(cls) -> 'type[TypeProto[In1T_contra, Out1T]]':
        ...
    @classmethod
    def push(cls, value: Out1T) -> None:
        ...

# This should (and does) match TypeProto[str, int].
class Int:
    @classmethod
    def convert(cls, value: str) -> int:
        return ... #type: ignore
    def resolve_item(self) -> Self:
        return ... #type: ignore
    @classmethod
    def resolve_type(cls) -> type[Self]:
        return ... #type: ignore
    @classmethod
    def push(cls, value: int) -> None:
        ...

# This second protocol is used to match against List[ElemT] and extract the underlying TypeProto signature of ElemT.
class ListProto(Protocol[In2T_contra, Out2T]):
    @property
    def element_type(self) -> TypeProto[In2T_contra, Out2T]:
        ...

# This is should (but doesn't with pyright) match TypeProto[Iterable[InT], list[OutT]] where ElemT matches TypeProto[InT, OutT].
class List(Generic[ElemT]):
    @property
    def element_type(self) -> ElemT:
        return ... #type: ignore
    @classmethod
    def convert(cls: type[ListProto[In3T, Out3T]], value: Iterable[In3T]) -> list[Out3T]:
        return ... #type: ignore
    def resolve_item(self) -> Self:
        return ... #type: ignore
    @classmethod
    def resolve_type(cls) -> type[Self]:
        return ... #type: ignore
    @classmethod
    def push(cls: type[ListProto[Never, Out3T]], value: list[Out3T]) -> None:  # I'm not 100% sure if this should use Never, but it doesn't seem to make a difference if I use eg Any instead.
        ...   

# Dummy function used to see if matching is working and to extract the input/output types.
def f(x: TypeProto[In4T, Out4T]) -> tuple[In4T, Out4T]:
    return ... #type: ignore

reveal_type(Int.convert)                # Type of "Int.convert" is "(value: str) -> int"
reveal_type(Int.resolve_item)           # Type of "Int.resolve_item" is "(self: Int) -> Int"
reveal_type(Int.resolve_type)           # Type of "Int.resolve_type" is "() -> type[Int]"
reveal_type(Int.push)                   # Type of "Int.push" is "(value: int) -> None"
reveal_type(f(Int()))                   # Type of "f(Int())" is "tuple[str, int]"

reveal_type(List[Int]().element_type)   # Type of "List[Int]().element_type" is "Int"
reveal_type(List[Int].convert)          # Type of "List[Int].convert" is "(value: Iterable[str]) -> list[int]"
reveal_type(List[Int].resolve_item)     # Type of "List[Int].resolve_item" is "(self: List[Int]) -> List[Int]"
reveal_type(List[Int].resolve_type)     # Type of "List[Int].resolve_type" is "() -> type[List[Int]]"
reveal_type(List[Int].push)             # Type of "List[Int].push" is "(value: list[int]) -> None"
reveal_type(f(List[Int]()))             # !!! Type of "f(List[Int]())" is "tuple[Unknown, Unknown]" !!!
```

As can be seen on the final line, pyright is unable to match `List[Int]` to the protocol despite all the relevant methods seeming to correctly match what is expected. As mentioned before, mypy is correctly matching this, leading me to suspect a bug in pyright.

```
note: Revealed type is "tuple[typing.Iterable[builtins.str], builtins.list[builtins.int]]"
```

Pyright reports the following error on that line:

```
error: Argument of type "List[Int]" cannot be assigned to parameter "x" of type "TypeProto[In4T@f, Out4T@f]" in function "f"
    "List[Int]" is incompatible with protocol "TypeProto[In4T@f, Out4T@f]"
      "resolve_item" is an incompatible type
        Type "() -> List[Int]" cannot be assigned to type "() -> TypeProto[In1T_contra@TypeProto, Out1T@TypeProto]"
          Function return type "List[Int]" is incompatible with type "TypeProto[In1T_contra@TypeProto, Out1T@TypeProto]"
            "List[Int]" is incompatible with protocol "TypeProto[In1T_contra@TypeProto, Out1T@TypeProto]"
      "resolve_type" is an incompatible type
        Type "() -> type[List[Int]]" cannot be assigned to type "() -> type[TypeProto[In1T_contra@TypeProto, Out1T@TypeProto]]"
          Function return type "type[List[Int]]" is incompatible with type "type[TypeProto[In1T_contra@TypeProto, Out1T@TypeProto]]"
    ... (reportGeneralTypeIssues)
```

But skipping everything in the middle it seems to boil down to something along the lines of "List[Int]" is incompatible with protocol "TypeProto[In4T@f, Out4T@f]" ... because ... "List[Int]" is incompatible with protocol "TypeProto[In1T_contra@TypeProto, Out1T@TypeProto]". It seems like somehow the recursive nature of the definitions of `resolve_item()` and `resolve_type()` is causing the issue. Removing those two functions from `TypeProto` does make pyright work, but I'm fairly sure there's nothing unsafe/incorrect about them.

```
# With resolve_item() and resolve_type() removed from TypeProto.
Type of "f(List[Int]())" is "tuple[Iterable[str], list[int]]"
```

Having said that, I did notice a curiosity that if instead of removing `resolve_item()` and `resolve_type()`, I instead remove the `push()` function, pyright also passes without error, however the revealed type on the final line isn't fully specified and contains the type variable `Out3T`:

```
# With resolve_item() and resolve_type(), but without the push() function.
Type of "f(List[Int]())" is "tuple[Iterable[str], list[int] | list[Out3T@convert]]"
```

None of the revealed types for the classmethods contained this (even when `push()` is removed from the protocol), so I'm not sure where it's coming from, and it's not mentioned in the original error message, but perhaps this is somehow getting in the way of the correct matching.

**VS Code extension or command-line**
I've tested this on the command line with `pyright 1.1.331`.

**Comment by erictraut** [comment]
I've fixed a couple of issues in response to your bug report.

First, I fixed a bug that led to a confusing error message in the case where protocol matching failed because an instance or class method could not be bound to the class. This was occurring in your code sample because of an underlying bug that prevented binding from being performed.

Second, I addressed the underlying cause of the false positive error. Pyright has an internal recursion limit for protocol matching. It's designed to protect against pathological cases (for performance and to prevent stack overflows in extreme cases). Your code was hitting this limit. I've updated the code to allow for a bit more breathing room.

This fix will be included in the next release.
